src/autoscaler/metricsforwarder/server/custom_metrics_handlers.go

```diff
@@ -0,0 +1,184 @@
+package server
+
+import (
+	"autoscaler/db"
+	"autoscaler/metricsforwarder/forwarder"
+	"autoscaler/models"
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"code.cloudfoundry.org/cfhttp/handlers"
+	"code.cloudfoundry.org/lager"
+	"github.com/patrickmn/go-cache"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type CustomMetricsHandler struct {
+	metricForwarder forwarder.MetricForwarder
+	policyDB        db.PolicyDB
+	credentialCache cache.Cache
+	cacheTTL        time.Duration
+	logger          lager.Logger
+}
+
+func NewCustomMetricsHandler(logger lager.Logger, metricForwarder forwarder.MetricForwarder, policyDB db.PolicyDB, credentialCache cache.Cache, cacheTTL time.Duration) *CustomMetricsHandler {
+	return &CustomMetricsHandler{
+		metricForwarder: metricForwarder,
+		policyDB:        policyDB,
+		credentialCache: credentialCache,
+		cacheTTL:        cacheTTL,
+		logger:          logger,
+	}
+}
+
+func (mh *CustomMetricsHandler) PublishMetrics(w http.ResponseWriter, r *http.Request, vars map[string]string) {
+	w.Header().Set("Content-Type", "application/json")
+	auth := strings.SplitN(r.Header.Get("Authorization"), " ", 2)
+	if len(auth) != 2 || auth[0] != "Basic" {
+		http.Error(w, "Authorization failed", http.StatusUnauthorized)
+		handlers.WriteJSONResponse(w, http.StatusUnauthorized, models.ErrorResponse{
+			Code:    "Authorization-Failure-Error",
+			Message: "Error varifying user credentials. Basic authorization is not used properly"})
+		return
+	}
+	payload, err := base64.StdEncoding.DecodeString(auth[1]) // Decoding the username:password
+
+	if err != nil {
+		mh.logger.Error("error-decoding-authorizaion-header", err, lager.Data{"authheader": auth})
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error decoding credentials"})
+		return
+	}
+
+	appID := vars["appid"]
+
+	pair := strings.SplitN(string(payload), ":", 2)
+
+	if len(pair) != 2 {
+		mh.logger.Error("error-processing-authorizaion-header", err, lager.Data{"authheader": auth})
+		http.Error(w, "Authorization failed", http.StatusUnauthorized)
+		handlers.WriteJSONResponse(w, http.StatusUnauthorized, models.ErrorResponse{
+			Code:    "Authorization-Failure-Error",
+			Message: "Basic authorization credential is not used properly"})
+		return
+	}
+
+	username := pair[0]
+	password := pair[1]
+	var isValid bool
+	var credentials models.CustomMetricCredentials
```

Comment:
if we put the hashed "username:password" directly in credential database, we won't need to handle username and password separately for each request.   In service instance 's credential information, the username is meaningless in fact.

Reply:
As discussed during the Credential Schema design phase, we will have username and password separate. As this proposal will not have much impact in general and the schema is already finalised I would like to skip this.